src/MallardMessageHandlers/AuthenticationToken/ConcurrentAuthenticationTokenProvider.cs

```diff
@@ -0,0 +1,135 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace MallardMessageHandlers
+{
+	/// <summary>
+	/// This proxy ensures that only 1 RefreshToken operation runs a any time.
+	/// It also ensures that only 1 NotifySessionExpired operation runs for equivalent expirations.
+	/// </summary>
+	/// <typeparam name="TAuthenticationToken">Type of authentication token</typeparam>
+	public class ConcurrentAuthenticationTokenProvider<TAuthenticationToken> : IAuthenticationTokenProvider<TAuthenticationToken>
+		where TAuthenticationToken : IAuthenticationToken
+	{
+		private readonly Func<CancellationToken, HttpRequestMessage, Task<TAuthenticationToken>> _getToken;
+		private readonly Func<CancellationToken, HttpRequestMessage, TAuthenticationToken, Task> _notifySessionExpired;
+		private readonly Func<CancellationToken, HttpRequestMessage, TAuthenticationToken, Task<TAuthenticationToken>> _refreshToken;
+
+		private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1, 1);
+		private readonly ILogger _logger;
+
+		private TAuthenticationToken _lastUnauthorizedToken;
```

Comment:
I'm not sure if keeping the last unauthorizedtoken is safe, since it can contains either personal info or the refresh token. That's why in my other version I kept only the token as a string.

Reply:
This isn't persisted on the phone.
Also, there's no real difference between the string or an object containing the string.